cmd/geth/dbcmd.go

```diff
@@ -759,3 +779,91 @@ func showMetaData(ctx *cli.Context) error {
 	table.Render()
 	return nil
 }
+
+func inspectHistory(ctx *cli.Context) error {
+	if ctx.NArg() == 0 || ctx.NArg() > 2 {
+		return fmt.Errorf("required arguments: %v", ctx.Command.ArgsUsage)
+	}
+	var (
+		address common.Address
+		slot    common.Hash
+	)
+	if err := address.UnmarshalText([]byte(ctx.Args().Get(0))); err != nil {
+		return err
+	}
+	if ctx.NArg() > 1 {
+		if err := slot.UnmarshalText([]byte(ctx.Args().Get(1))); err != nil {
+			return err
+		}
+	}
+	// Load the databases.
+	stack, _ := makeConfigNode(ctx)
+	defer stack.Close()
+
+	db := utils.MakeChainDatabase(ctx, stack, true)
+	defer db.Close()
+
+	triedb := utils.MakeTrieDatabase(ctx, db, false, false, false)
+	defer triedb.Close()
+
+	var (
+		err   error
+		start uint64 // the first history object to query
+		end   uint64 // the last history object to query (included)
+
+		// State histories are sorted by state ID rather than block number.
+		// To address this, load the corresponding block header and perform
+		// the conversion by this function.
+		blockToID = func(blockNumber uint64) (uint64, error) {
+			header := rawdb.ReadHeader(db, rawdb.ReadCanonicalHash(db, blockNumber), blockNumber)
+			if header == nil {
+				return 0, fmt.Errorf("block #%d is not existent", blockNumber)
+			}
+			id := rawdb.ReadStateID(db, header.Root)
+			if id == nil {
+				first, last, err := triedb.HistoryRange()
+				if err == nil {
+					return 0, fmt.Errorf("history of block #%d is not existent, available history range: [#%d-#%d]", blockNumber, first, last)
+				}
+				return 0, fmt.Errorf("history of block #%d is not existent", blockNumber)
+			}
+			return *id, nil
+		}
+	)
+	startNumber := ctx.Uint64("start")
+	if startNumber != 0 {
+		start, err = blockToID(startNumber)
+		if err != nil {
+			return err
+		}
+	}
+	endBlock := ctx.Uint64("end")
+	if endBlock != 0 {
+		end, err = blockToID(endBlock)
+		if err != nil {
+			return err
+		}
+	}
+	var stats *pathdb.HistoryStats
+	if slot == (common.Hash{}) {
+		stats, err = triedb.AccountHistory(address, start, end)
+	} else {
+		// The hash of storage slot key is utilized in the history
+		// rather than the raw slot key, make the conversion.
+		slotHash := crypto.Keccak256Hash(slot.Bytes())
+		stats, err = triedb.StorageHistory(address, slotHash, start, end)
+	}
+	if err != nil {
+		return err
+	}
+	if slot == (common.Hash{}) {
+		fmt.Printf("Account history: %s\n", address.Hex())
+	} else {
+		fmt.Printf("Storage history: %s-%s\n", address.Hex(), slot.Hex())
+	}
+	fmt.Printf("Range: [#%d-#%d]\n\n", stats.Start, stats.End)
+	for i := 0; i < len(stats.Blocks); i++ {
+		fmt.Printf("#%d: %x\n", stats.Blocks[i], stats.Origins[i])
```

Comment:
Just to be clear, if the output says
```
#4973565: 8902f2b4e4eb60ea5800
```
That means that the slot was set to value `0x8902f2b4e4eb60ea5800` within block `4973565`, yes? It's not like "this was the previous value until `4973565`" ? 

Also perhaps 
```suggestion
		fmt.Printf("#%d: %#x\n", stats.Blocks[i], stats.Origins[i])
```
?

Reply:
the value means: `"this was the previous value until 4973565"`

Reply:
I see. Can we make that clearer? Maybe just print out a comment in the beginning
```
Storage history:
 account: xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2
 slot: 0x321690423bfa8088917640d62a3b42bb0acc95376d9366929f697905e89aead8
 blockrange: [#4972612-#4974412]

The output shows block number where slot was modified, along with the _old_ value.

#4972815: 88180a84df00995800
#4973208: 89027aa9ebbfc5c95800
#4973565: 8902f2b4e4eb60ea5800
```

Reply:
Will do!

Reply:
done